src/app/websites/website-services/afterdark/afterdark.service.ts

```diff
@@ -0,0 +1,201 @@
+  if (collections.length < 1) {
+    return [['AfterDark requires at least one collection folder', { website: 'AfterDark', value: collections.length }]];
+  }
+  return [];
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+@Website({
+  acceptedFiles: ACCEPTED_FILES,
+  additionalFiles: false,
+  displayedName: 'AfterDark',
+  login: {
+    url: 'https://afterdark.art/login',
+  },
+  components: {
+    submissionForm: AfterDarkSubmissionForm,
+  },
+  validators: {
+    submission: submissionValidate
+  },
+  preparsers: {},
+  parsers: {
+    description: [(html: string) => html],
+  },
+})
+export class AfterDark extends BaseWebsiteService implements WebsiteService {
+  readonly BASE_URL: string = 'https://afterdark.art';
+
+  constructor(private _profileManager: LoginProfileManagerService) {
+    super();
+  }
+
+  private async getCsrfToken(profileId: string): Promise<string> {
+    var cookies = await getCookies(profileId, this.BASE_URL);
+    const response = await got.get(this.BASE_URL, this.BASE_URL, cookies, profileId);
+    if (response.statusCode !== 200) {
+      throw new Error(`Bad response getting csrf token ${response.statusCode}`);
+    }
+    const match = response.body.match(new RegExp(`name="csrfmiddlewaretoken" value="([^"]+)"`));
+    if (!match) {
+      throw new Error('Could not find csrf token in body');
+    }
+    const csrfToken = match[1];
+    return csrfToken;
+  }
+
+  private async getUserData(profileId: string, csrfToken: string): Promise<any> {
+    const body = JSON.stringify({
+      "operationName": "UserQuery",
+      "variables": {},
+      "query": `query UserQuery {
+        userViewer {
+          user {
+            id
+            username
+            isUserAdmin
+            isProUser
+            isEmailVerified
+            collectionSet {
+              edges {
+                node {
+                  id
+                  title
+                }
+              }
+            }
+          }
+        }
+      }`
+    });
+    var cookies = await getCookies(profileId, this.BASE_URL);
+    const response = await got.gotPostJSON(`${this.BASE_URL}/graphql/`, body, this.BASE_URL, cookies, profileId, {
+      headers: {
+        'content-type': 'application/json',
+        'X-CSRFToken': csrfToken,
+      }
+    });
+    if (response.statusCode !== 200) {
+      throw new Error(`Bad response getting user data ${response.statusCode}`);
+    }
+    const responseBody = JSON.parse(response.body);
+    if (!(responseBody.data || {}).userViewer) {
+      return null;
+    }
+    const collections = ((((responseBody.data || {}).userViewer || {}).user || {}).collectionSet || {}).edges || [];
+    return {
+      'userId': (((responseBody.data || {}).userViewer || {}).user || {}).id || '',
+      'username': (((responseBody.data || {}).userViewer || {}).user || {}).username || '',
+      'collections': collections.map(c => c.node),
```

Comment:
All of this chaining is ugly and error prone.
Look at website-validatgor.helper.ts (getTags) as an example and use dotProp.get(...)

Reply:
This library is great!  Done.